fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipServiceImpl.java

```diff
@@ -0,0 +1,438 @@
+            indexManager.endMembershipForSource(txId, fedoraId, fedoraResc.getLastModifiedDate());
+        }
+
+        // delete child of DirectContainer, clear from tx and end existing
+        final var parentResc = getParentResource(fedoraResc);
+
+        if (isDirectContainer(parentResc)) {
+            final var parentRdfResc = getRdfResource(parentResc);
+            final var deletedMembership = generateDirectMembership(txId, parentRdfResc, fedoraResc);
+            indexManager.endMembership(txId, parentResc.getFedoraId(), deletedMembership,
+                    fedoraResc.getLastModifiedDate());
+
+        }
+    }
+
+    @Override
+    public RdfStream getMembership(final String txId, final FedoraId fedoraId) {
+        final FedoraId subjectId;
+        if (fedoraId.isDescription()) {
+            subjectId = fedoraId.asBaseId();
+        } else {
+            subjectId = fedoraId;
+        }
+        final var subject = NodeFactory.createURI(subjectId.getBaseId());
+        final var membershipStream = indexManager.getMembership(txId, subjectId);
+        return new DefaultRdfStream(subject, membershipStream);
+    }
+
+    @Override
+    public void commitTransaction(final String txId) {
+        indexManager.commitTransaction(txId);
+    }
+
+    @Override
+    public void rollbackTransaction(final String txId) {
+        indexManager.deleteTransaction(txId);
+    }
+
+    @Override
+    public void reset() {
+        indexManager.clearIndex();
+    }
+
+    @Override
+    @Transactional
+    public void populateMembershipHistory(final String txId, final FedoraId containerId) {
+        final FedoraResource fedoraResc = getFedoraResource(txId, containerId);
+
+        if (isDirectContainer(fedoraResc)) {
+            populateMembershipHistory(txId, fedoraResc, null);
+        }
+    }
+
+    private void populateMembershipHistory(final String txId, final FedoraResource fedoraResc,
+            final Instant afterTime) {
+        final var containerId = fedoraResc.getFedoraId();
+        final var propertyTimeline = makePropertyTimeline(fedoraResc);
+        final List<DirectContainerProperties> timeline;
+        // If provided, filter the timeline to just entries active on or after the specified time
+        if (afterTime != null) {
+            timeline = propertyTimeline.stream().filter(e -> e.startDatetime.compareTo(afterTime) >= 0
+                    || e.endDatetime.compareTo(afterTime) >= 0)
+                .collect(Collectors.toList());
+        } else {
+            timeline = propertyTimeline;
+        }
+
+        // get all the members of the DC and index the history for each, accounting for changes to the DC
+        fedoraResc.getChildren().forEach(member -> {
+            final var memberDeleted = member instanceof Tombstone;
+            final var memberNode = NodeFactory.createURI(member.getFedoraId().getFullId());
+            log.debug("Populating membership history for DirectContainer {}member {}",
```

Comment:
Maybe add a `,` between "{}" and "member"?

Reply:
I know it looks goofy, but that would result in two spaces between `DirectContainer` and `member` if the member isn't deleted. I could remove the deletion state from the debug statement though